// ngx-image-zoom/esm2020/lib/ngx-image-zoom.module.mjs
import { NgModule } from '@angular/core';
import { NgxImageZoomComponent } from './ngx-image-zoom.component';
import { CommonModule } from '@angular/common';
import * as i0 from "@angular/core";
export class NgxImageZoomModule {
}
NgxImageZoomModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.1.3", ngImport: i0, type: NgxImageZoomModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgxImageZoomModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.1.3", ngImport: i0, type: NgxImageZoomModule, declarations: [NgxImageZoomComponent], imports: [CommonModule], exports: [NgxImageZoomComponent] });
NgxImageZoomModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.1.3", ngImport: i0, type: NgxImageZoomModule, imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.1.3", ngImport: i0, type: NgxImageZoomModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        NgxImageZoomComponent,
                    ],
                    imports: [
                        CommonModule,
                    ],
                    exports: [
                        NgxImageZoomComponent
                    ]
                }]
        }] });
